File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/operator_oveloading/java/Money.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.operator_oveloading.java;

public class Money {

    private final double amount;
    private final String currency;

    public Money(double amount, String currency) {
        this.amount = amount;
        this.currency = currency;
    }

    public Money add(Money other) {
        // Java does not support operator overloading.
        // Domain operations must be represented as methods.
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException("Currencies must be the same");
        }

        return new Money(
                amount + other.amount,
                currency
        );
    }

    public double getAmount() {
        return amount;
    }

    public String getCurrency() {
        return currency;
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/operator_oveloading/kotlin/Money.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.operator_oveloading.kotlin

data class Money(
    val amount: Double,
    val currency: String
) {
    // Kotlin supports operator overloading.
    // This allows using the + operator for domain objects.
    operator fun plus(other: Money): Money {
        require(currency == other.currency) {
            "Currencies must be the same"
        }

        return Money(
            amount = amount + other.amount,
            currency = currency
        )
    }
}

class MoneyService {

    fun calculateTotal(): Money {
        val first = Money(100.0, "PLN")
        val second = Money(50.0, "PLN")

        // This calls the overloaded plus function.
        return first + second
    }
}